diff --git a/tracblockdiag/cache.py b/tracblockdiag/cache.py
index 8dede9e..6cc57f1 100644
--- a/tracblockdiag/cache.py
+++ b/tracblockdiag/cache.py
@@ -13,8 +13,8 @@ def is_obsolete(entry, duration):
 
 
 def compute_key(function, args, kwargs):
-    key = pickle.dumps((function.func_name, args, kwargs))
-    return hashlib.sha1(key).hexdigest()
+    key = pickle.dumps((function.__name__, args, kwargs))
+    return hashlib.sha256(key).hexdigest()
 
 
 def memoize(duration=300):
diff --git a/tracblockdiag/diag.py b/tracblockdiag/diag.py
index f8cb29a..0f86e16 100644
--- a/tracblockdiag/diag.py
+++ b/tracblockdiag/diag.py
@@ -4,10 +4,18 @@ import os
 import sys
 from threading import RLock
 
-try:
-    from cStringIO import StringIO
-except ImportError:
-    from StringIO import StringIO
+
+if 2 < sys.version_info.major:
+    try:
+        # for python3
+        from io import BytesIO
+    except ImportError:
+        pass
+else:
+    try:
+        from cStringIO import StringIO
+    except ImportError:
+        from StringIO import StringIO
 
 try:
     from blockdiag.utils.fontmap import FontMap
@@ -42,7 +50,11 @@ class BaseBuilder(object):
         return draw(diagram, options)
 
     def draw_png(self, diagram, options):
-        sio = StringIO()
+        sio = None
+        if 2 < sys.version_info.major:
+            sio = BytesIO()
+        else:
+            sio = StringIO()
         drawer = self.DiagramDraw('PNG', diagram, sio, **options)
         drawer.draw()
         drawer.save()
diff --git a/tracblockdiag/plugin.py b/tracblockdiag/plugin.py
index 952530c..0bbffb6 100644
--- a/tracblockdiag/plugin.py
+++ b/tracblockdiag/plugin.py
@@ -48,7 +48,7 @@ class BlockdiagRenderer(Component):
     def __init__(self):
         self.font = self.config.getlist(_conf_section, 'font')
         self.default_type = self.config.get(_conf_section, 'default_type',
-                                            'svg')
+                                            'png')
         self.fallback = self.config.getbool(_conf_section, 'fallback', False)
         self.syntax_check = self.config.getbool(_conf_section, 'syntax_check',
                                                 True)
@@ -105,7 +105,7 @@ class BlockdiagRenderer(Component):
         return html.object(**kwargs)
 
     def get_url(self, diag, type_, data):
-        return self.url_template % {'diag': diag, 'type': type_, 'data': data}
+        return self.url_template % {'diag': diag, 'type': type_, 'data': data.decode('utf-8')}
 
     def check_syntax(self, kind, content):
         try:
